src/insert.rs

```diff
@@ -185,14 +182,40 @@ impl<T> Insert<T> {
         self
     }
 
+    /// Set the [role] to use when executing `INSERT` statements.
+    ///
+    /// Overrides any role previously set by [`Client::with_role()`] or [`Client::with_option()`].
+    ///
+    /// [role]: https://clickhouse.com/docs/operations/access-rights#role-management
+    ///
+    /// # Panics
+    /// If called after the request is started, e.g., after [`Insert::write`].
+    pub fn with_role(mut self, role: impl Into<String>) -> Self {
+        self.state.expect_client_mut().set_role(Some(role.into()));
+        self
+    }
+
+    /// Perform inserts without any explicit [role] set.
+    ///
+    /// Overrides any role previously set by [`Client::with_role()`] or [`Client::with_option()`].
+    ///
+    /// [role]: https://clickhouse.com/docs/operations/access-rights#role-management
+    ///
+    /// # Panics
+    /// If called after the request is started, e.g., after [`Insert::write`].
+    pub fn with_default_role(mut self) -> Self {
+        self.state.expect_client_mut().set_role(None);
+        self
+    }
```

Comment:
Let's do only one method: `with_roles(impl IntoIterator<Item = impl Into<String>>)` as discussed. An empty slice means no roles are set via URL parameters. The behavior should be the same for Client, Query, and Insert, where Client roles are lower priority and can be overridden by Query or Insert method calls, very similarly to `with_option`.